Storage/DiskImageStore.swift

```diff
@@ -84,18 +84,22 @@ open class DiskImageStore {
     }
 
     /// Clears all images from the cache, excluding the given set of keys.
-    open func clearExcluding(_ keys: Set<String>) {
-        let keysToDelete = self.keys.subtracting(keys)
+    open func clearExcluding(_ keys: Set<String>) -> Success {
+        return deferDispatchAsync(queue) { () -> Success in
+            let keysToDelete = self.keys.subtracting(keys)
 
-        for key in keysToDelete {
-            let url = URL(fileURLWithPath: filesDir).appendingPathComponent(key)
-            do {
-                try FileManager.default.removeItem(at: url)
-            } catch {
-                log.warning("Failed to remove DiskImageStore item at \(url.absoluteString): \(error)")
+            for key in keysToDelete {
+                let url = URL(fileURLWithPath: self.filesDir).appendingPathComponent(key)
+                do {
+                    try FileManager.default.removeItem(at: url)
```

Comment:
So you now have a serial queue — one of these callbacks can run at a time — and you're doing an unbounded number of one-by-one file deletions that blocks the queue.

This is a somewhat risky change: you've limited the throughput of the system dramatically in order to avoid a concurrency issue, and that might cause `UIImage` instances to stick around for a long time.